### ISSUE TITLE
Broker failed with error: Group completed error: redigo: nil returned

### ISSUE THREAD

**Issue by gow** [RichardKnop/machinery#352]
Hi,
We recently started to use Machinery to run some of our asynchronous jobs. I've noticed that a couple of times the worker abruptly stopped processing tasks. I found the following entry in the logs which was logged at the same time when I noticed the outage. But it doesn't tell much.

`Broker failed with error: Group completed error: redigo: nil returned`

Looking into the code, this log is emitted when the worker is trying to decide if the group is complete by fetching the status of all tasks within. I didn't notice any errors in my redis server either. 

1) Has anyone come across this issue before? How did you fix it?
2) Is this failure to check group completion status related to workers not processing any tasks?

Any pointers would be greatly appreciated.

**Comment by gow** [comment]
Ok. After debugging further, I discovered a couple of issues that might be related to this.

**1. The group result TTL is not set correctly.**
**Issue**: I had the default result expiry time (1 hour) and had tasks execute in a chord. This set the group's result to expire in 1hr when it was created. But some tasks in the group took longer than 1 hour to complete because of retries and exponential backoff. So, by the time all the individual tasks finished, the group's result had expired and the framework didn't know how to proceed and the chord was not executed completely.
**Fix**: Set the group's expiration time to Max(tasks.ETA) + ResultsExpiryTime

**2. The above issue somehow causes the workers to drop other tasks from the queue.**
**Issue**: While testing locally, I noticed that when the worker encountered the `Broker failed with error: Group completed error: redigo: nil returned` error, it dropped all other tasks currently existing in the queue. It did process newly enqueued jobs though. I think this is a bigger issue.
**Fix**: Don't know

**3. Wrap the errors with useful information.**
**Issue**: The errors found in the logs were not descriptive. 
**Fix**: Add useful information like the Group ID that was not found etc.

**Comment by feelinc** [comment]
I have similar issue `worker.go:69 Broker failed with error: Group completed error: redigo: nil returned`

Broker : SQS
Backend : Redis
Queue : FIFO

This happen if I set `GroupUUID` in task.

**Comment by risav1994** [comment]
It can also happen if you have set two different db number for two different tasks in a group.

**Comment by RichardKnop** [comment]
@gow This is probably caused by default expiration time being 1 hour: https://github.com/RichardKnop/machinery/blob/master/v1/config/config.go#L19

Please set it to larger value otherwise task states and group metadata will be expired. I am going to push a commit to make 24 hours default expiry time as 1 hour seems quite short.

For the issue 2) I am not sure why that would happen, could you debug more and find out the reason? I Don't think any tasks should be lost from the queue as there is no expiration time set for messages in Redis queue Redis broker consumes from.

**Comment by gow** [comment]
@RichardKnop yes. Setting the default timeout to 24hr could provide some relief. I changed our config to 24hrs as well. But someone will definitely run into this problem again at some point. All they have to do is to have retries + backoff configured such that it the task/group exists longer than the configured `ResultsExpiryTime`. I think the right fix is to extend the TTL of the result object whenever a task is put back into the queue due to a retry.

Regarding issue 2): I too found it strange. But something to that effect did happen. I'll investigate it further when I get some time.

While you are at it, can you also address issue-3?
A more descriptive error message would have saved a bit of time and also pointed me towards the right issue sooner. Something like `Broker failed with error: Group [Group-ID] completed error: redigo: nil returned` would be more helpful. Take a look at https://github.com/pkg/errors. It makes it easier to wrap errors. I'll be happy to help and send a PR as well.

**Comment by bp-fndr** [comment]
@feelinc  I believe you are running into this issue I opened here: https://github.com/RichardKnop/machinery/issues/364
